jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java

```diff
@@ -210,11 +211,72 @@ public void clear()
     @Override
     public String toString()
     {
-        return String.format("%s@%x{i=%s,ip=%s,e=%s,ep=%s}",this.getClass().getSimpleName(),hashCode(),_includes,_includePredicate,_excludes,_excludePredicate);
+        return String.format("%s@%x{i=%s,ip=%s,e=%s,ep=%s}",this.getClass().getSimpleName(),hashCode(),
+            _includes,
+            _includePredicate==_includes?"SELF":_includePredicate,
+            _excludes,
+            _excludePredicate==_excludes?"SELF":_excludePredicate);
     }
 
     public boolean isEmpty()
     {
         return _includes.isEmpty() && _excludes.isEmpty();
     }
+
+
+    /**
+     * Combine the results of two {@link IncludeExcludeSet}s with an "and" operation.  The items must be
+     * included by xSet AND ySet, and excluded from neither.   If a sets inclusions are empty, then all items
+     * are considered to be included.
+     * @param xSet The set that param x is tested against
+     * @param x An item to test against xSet
+     * @param ySet The set that param y is tested against
+     * @param y A supplier of an item to test against ySet, that is executed only of x is not excluded from xSet
+     * @param <XS> The type of xSet items
+     * @param <XP> The type of xSet predicates
+     * @param <YS> The type of ySet items
+     * @param <YP> The type of ySet predicates
+     * @return True only if the items are included and not excluded from their respected sets
+     */
+    public static <XS, XP, YS, YP> boolean and(IncludeExcludeSet<XS, XP> xSet, XP x, IncludeExcludeSet<YS,YP> ySet, Supplier<YP> y)
+    {
+        Boolean xb = xSet.isIncludedAndNotExcluded(x);
+        if (Boolean.FALSE==xb || (xb==null && xSet.hasIncludes()))
+            return false;
+        Boolean yb = ySet.isIncludedAndNotExcluded(y.get());
+        if (Boolean.FALSE==yb || (yb==null && ySet.hasIncludes()))
+            return false;
+
+        return true;
+    }
+
+
+    /**
+     * Combine the results of two {@link IncludeExcludeSet}s with an "or" operation.  The items must be
+     * included by xSet OR ySet, and excluded from neither.   If a sets inclusions are empty, then all items
+     * are considered to be included.
+     * @param xSet The set that param x is tested against
+     * @param x An item to test against xSet
+     * @param ySet The set that param y is tested against
+     * @param ySupplier A supplier of an item to test against ySet, that is executed only of x is not excluded from xSet
+     * @param <XS> The type of xSet items
+     * @param <XP> The type of xSet predicates
+     * @param <YS> The type of ySet items
+     * @param <YP> The type of ySet predicates
+     * @return True only if the items are included and not excluded from their respected sets
+     */
+    public static <XS, XP, YS, YP> boolean or(IncludeExcludeSet<XS, XP> xSet, XP x, IncludeExcludeSet<YS,YP> ySet, Supplier<YP> ySupplier)
```

Comment:
Ditto for this one.  why is `Predicate.or` not viable here?

Reply:
because Predicate.or only works if the type of the two predicates is the same, which in this case it is not.

Reply:
Too difficult to understand. See above for simpler alternative.

Reply:
Perhaps the and logic is not necessary, but the or logic is necessary.    If one set has an include, that changes how the other set is considered.    For example with the classloader if a the package name set just has excludes, but the location set as some includes, then it is not sufficient to say  matches packages or matches locations.  The presence of an include in the location set switches both sets to require positive inclusion rather than just not exclusion.

Reply:
Review this - should not be `combine()`?